### PR TITLE
Multicall3 Final Review

### PR DESCRIPTION
**Pull request by mds1**
## Overview

The main file to review is `Multicall3.sol`. It's ABI is backwards compatible with Multicall and Multicall2, but it's cheaper to use (so you can fit more calls into a single request), and it adds an `aggregate3` method so you can specify whether calls are allowed to fail on a per-call basis. Additionally, it will be deployed at every network on the same address.

Any reviews/feedback on that contract (and optionally the rest of this PR) prior to deployment is appreciated!

## Deliverables

- [x] Merge #4 
- [x] Add `aggregate3Value` method that takes a `Call3Value` struct, for sending value when used on-chain
- [x] Merge #6 
  - [x] Add `README.md` contract documentation and gas-golfing explainers
  - [x] Verify this saves gas and implement it https://github.com/mds1/multicall/pull/2#discussion_r817325023
  - [x] Remove `blockNumber` and `blockHash` from `aggregate3` as per https://github.com/mds1/multicall/pull/2#discussion_r817323516
  - [x] Implement https://github.com/mds1/multicall/pull/2#discussion_r818725378
  - [x] Optimize call value pointer https://github.com/mds1/multicall/pull/2#discussion_r820051073
  - [x] Add `getChainId` and `getBasefee` accessors

### CODE REVIEW

File: src/Multicall3.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    function blockAndAggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    function aggregate3(Call3[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
```
Comment (mds1) [inline src/Multicall3.sol:62]: Should we change this to only return `Result[] memory returnData`, instead of always returning `blockNumber` and `blockHash`? The rationale for doing so is that instead of being forced to return block number and hash, you can choose which block parameters to return (e.g. maybe you only care about timestamp) by simply including them as calls in the Multicall

File: src/Multicall3.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(calls[i].allowFailure || success, "Multicall3: aggregate3 failed");
```
Comment (mds1) [inline src/Multicall3.sol:69]: @transmissions11 mentioned we can save gas in this check by instead using assembly's `or()` operator to avoid short-circuiting. This would remove a JUMPI which costs 10 gas, so just would need to verify that 10 gas is more than the cost to evaluate the right-hand side of the expression

**Comment by joshieDo** [comment]
Caching at `aggregate3` seems to have some nice gas savings, unless foundry is lying to me:

```
Call3 calldata call = calls[i];
```

File: src/Multicall3.sol (hunk 2 above)
Comment (ind-igo) [reply src/Multicall3.sol:69]: Should this also use a custom error instead of an error string?

**Comment by abigger87** [comment]
Any references to calldata pointers/quick explanation of how they reduce gas? cc @joshieDo @transmissions11 ? Otherwise I can always pour over the asm

File: src/Multicall3.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function aggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall3: aggregate failed");
+            returnData[i] = ret;
```
Comment (hrkrshnn) [inline src/Multicall3.sol:36]:
```suggestion
            bool success;
            (success, returnData[i]) = calls[i].target.call(calls[i].callData);
            require(success, "Multicall3: aggregate failed");
```
That should save an allocation.

File: src/Multicall3.sol (hunk 2 above)
Comment (hrkrshnn) [reply src/Multicall3.sol:69]: > This would remove a JUMPI which costs 10 gas, so just would need to verify that 10 gas is more than the cost to evaluate the right-hand side of the expression

This is very likely the case.

**Comment by abigger87** [comment]
> Caching at aggregate3 seems to have some nice gas savings, unless foundry is lying to me:
> Edit: also works with caching Call

Added in #3 
Great suggestion @joshieDo !

File: src/Multicall3.sol (hunk 2 above)
Comment (abigger87) [reply src/Multicall3.sol:69]: > Should this also use a custom error instead of an error string?

@ind-igo If it were my preference, we would, but using <=32 byte strings for backwards-compatibility. We're trying to optimize wrt to the tradeoff of backwards-compatibility <> optimizing.

**Comment by odyslam** [comment]
Why do you pass blockNumber as an argument and then assign right away `block.number`?

File: src/Multicall3.sol (hunk 2 above)
Comment (odyslam) [reply src/Multicall3.sol:69]: Would it also make sense to have `success` first in the evaluation, assuming that the user knows what they do so it has a higher chance to be correct and stop EVM from evaluating the mapping

File: src/Multicall3.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.12;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
```
Comment (fqlx) [inline src/Multicall3.sol:5]: Doesn't multicall support writing function calls?

File: src/Multicall3.sol (hunk 3 above)
Comment (mds1) [reply src/Multicall3.sol]: 75f24aa changes this and adds calldata pointers—thanks @hrkrshnn and @joshieDo!

File: src/Multicall3.sol (hunk 2 above)
Comment (mds1) [reply src/Multicall3.sol]: Hmm, yea I wonder which condition would be true more often. I'm inclined to think `success` meaning it should be first. But either way it sounds like using `or()` to remove a JUMPI will be cheapest, I just have not actually implemented and benchmarked it yet (maybe @abigger87 has?)

**Comment by mds1** [comment]
Sorry @odyslam what line are you referring to? Could you leave a comment on that line instead?

File: src/Multicall3.sol (hunk 4 above)
Comment (mds1) [reply src/Multicall3.sol:5]: True, it does! I think read-only is the more common use case so how about
```suggestion
/// @notice Aggregate results from multiple read-only function calls or batch on-chain calls into a single transaction
```

File: src/Multicall3.sol (hunk 2 above)
Comment (odyslam) [reply src/Multicall3.sol]: > Hmm, yea I wonder which condition would be true more often. I'm inclined to think `success` meaning it should be first. But either way it sounds like using `or()` to remove a JUMPI will be cheapest, I just have not actually implemented and benchmarked it yet (maybe @abigger87 has?)

@mds1 I read about short-circuiting, but i can't grep why the assembly or will save a JUMPI. ELI5 pls ser 🥲

Comment (hrkrshnn) [reply src/Multicall3.sol]: @odyslam A solidity high level `if (a && b) {...}` gets converted to the following in Yul. 

```
if a {
    if b {
        ....
    }
}
```

This allows for short-circuiting. Assume that the expressions `a` and `b` are state changing, if `a` returns false, then `b` would never be executed. But if the expressions `a` and `b` are side effect free, then `or(a, b)` may be cheaper.
```
if or(a, b) {
    ....
}
```

**Comment by joshieDo** [comment]
make every public non-view `payable` ?

```diff
diff --git a/.gas-snapshot b/.gas-snapshot
index 3593ce3..98b0c1c 100644
--- a/.gas-snapshot
+++ b/.gas-snapshot
@@ -13,10 +13,10 @@ Multicall2Test:testTryAggregateUnsuccessful() (gas: 10212)
 Multicall2Test:testTryBlockAndAggregate() (gas: 10671)
 Multicall2Test:testTryBlockAndAggregateUnsuccessful() (gas: 10360)
 Multicall2Test:testUnsuccessulAggregation() (gas: 9852)
-Multicall3Test:testAggregate3() (gas: 10561)
-Multicall3Test:testAggregate3Unsuccessful() (gas: 10258)
-Multicall3Test:testAggregation() (gas: 5797)
-Multicall3Test:testBlockAndAggregateUnsuccessful() (gas: 9875)
+Multicall3Test:testAggregate3() (gas: 10537)
+Multicall3Test:testAggregate3Unsuccessful() (gas: 10234)
+Multicall3Test:testAggregation() (gas: 5773)
+Multicall3Test:testBlockAndAggregateUnsuccessful() (gas: 9851)
 Multicall3Test:testGetBlockHash(uint256) (runs: 1000, μ: 1196, ~: 1196)
 Multicall3Test:testGetBlockNumber() (gas: 942)
 Multicall3Test:testGetCurrentBlockCoinbase() (gas: 1034)
@@ -25,11 +25,11 @@ Multicall3Test:testGetCurrentBlockGasLimit() (gas: 963)
 Multicall3Test:testGetCurrentBlockTimestamp() (gas: 976)
 Multicall3Test:testGetEthBalance(address) (runs: 1000, μ: 3944, ~: 3944)
 Multicall3Test:testGetLastBlockHash() (gas: 270)
-Multicall3Test:testTryAggregate() (gas: 9807)
-Multicall3Test:testTryAggregateUnsuccessful() (gas: 9836)
-Multicall3Test:testTryBlockAndAggregate() (gas: 10154)
-Multicall3Test:testTryBlockAndAggregateUnsuccessful() (gas: 9939)
-Multicall3Test:testUnsuccessulAggregation() (gas: 9534)
+Multicall3Test:testTryAggregate() (gas: 9783)
+Multicall3Test:testTryAggregateUnsuccessful() (gas: 9812)
+Multicall3Test:testTryBlockAndAggregate() (gas: 10130)
+Multicall3Test:testTryBlockAndAggregateUnsuccessful() (gas: 9915)
+Multicall3Test:testUnsuccessulAggregation() (gas: 9510)
 MulticallTest:testAggregation() (gas: 5899)
 MulticallTest:testGetBlockHash(uint256) (runs: 1000, μ: 1173, ~: 1173)
 MulticallTest:testGetCurrentBlockCoinbase() (gas: 1034)
```

File: src/Multicall3.sol [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    function getEthBalance(address addr) public view returns (uint256 balance) {
+        balance = addr.balance;
+    }
+
+    function getLastBlockHash() public view returns (bytes32 blockHash) {
+        unchecked {
+            blockHash = blockhash(block.number - 1);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mds1) [reply src/Multicall3.sol:114]: Dispatch for `aggregate3` only adds about 70 gas, so I think we can leave this as-is and not handle dispatching ourselves

**Comment by mds1** [comment]
Here's what I think is left before merge/deploy:

- [ ] merge #4 
- [ ] add `aggregate3Value` method that takes a `Call3Value` struct, for sending value when used on-chain
- [ ] add comment at the top for payable / mentioning the contract is gas-golfed
- [ ] verify this saves gas and implement it https://github.com/mds1/multicall/pull/2#discussion_r817325023
- [ ] decision on https://github.com/mds1/multicall/pull/2#discussion_r817323516
- [ ] implement https://github.com/mds1/multicall/pull/2#discussion_r818725378

File: src/Multicall3.sol (hunk 1 above)
Comment (abigger87) [reply src/Multicall3.sol]: Leaning towards removing `blockNumber` and `blockHash` although it would follow Multicall and Multicall2 canonical patterns.

For example, snapshot.js ignores the blockNumber for `multicall` https://github.com/abigger87/snapshot.js/blob/master/src/utils.ts#L122

File: src/Multicall3.sol [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        Call3Value calldata call;
+        for (uint256 i = 0; i < length;) {
+            Result memory result = returnData[i];
+            call = calls[i];
+            // Humanity will be a Type V Kardashev Civilization before this overflows - andreas
+            // ~ 10^25 Wei in existence << ~ 10^76 size uint fits in a uint256
+            unchecked { valAccumulator += call.value; }
+            (result.success, result.returnData) = call.target.call{value: call.value}(call.callData);
```
Comment (joshieDo) [inline src/Multicall3.sol:129]:
```suggestion
            uint256 val = call.value;
            // Humanity will be a Type V Kardashev Civilization before this overflows - andreas
            // ~ 10^25 Wei in existence << ~ 10^76 size uint fits in a uint256
            unchecked { valAccumulator += val; }
            (result.success, result.returnData) = call.target.call{value: val}(call.callData);
```
Saves accessing the value again.

```diff
diff --git a/.gas-snapshot b/.gas-snapshot
index def28f3..9fb01fb 100644
--- a/.gas-snapshot
+++ b/.gas-snapshot
@@ -15,8 +15,8 @@ Multicall2Test:testTryBlockAndAggregateUnsuccessful() (gas: 10382)
 Multicall3Test:testAggregate3Unsuccessful() (gas: 10279)
-Multicall3Test:testAggregate3Value() (gas: 34757)
-Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62666)
+Multicall3Test:testAggregate3Value() (gas: 34736)
+Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62608)
 Multicall3Test:testAggregation() (gas: 5883)
```

Comment (joshieDo) [inline src/Multicall3.sol:129]:
```suggestion
            uint256 val = call.value;
            // Humanity will be a Type V Kardashev Civilization before this overflows - andreas
            // ~ 10^25 Wei in existence << ~ 10^76 size uint fits in a uint256
            unchecked { valAccumulator += val; }
            (result.success, result.returnData) = call.target.call{value: val}(call.callData);
```

```diff
diff --git a/.gas-snapshot b/.gas-snapshot
index def28f3..9fb01fb 100644
--- a/.gas-snapshot
+++ b/.gas-snapshot
@@ -15,8 +15,8 @@ Multicall2Test:testTryBlockAndAggregateUnsuccessful() (gas: 10382)
 Multicall3Test:testAggregate3Unsuccessful() (gas: 10279)
-Multicall3Test:testAggregate3Value() (gas: 34757)
-Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62666)
+Multicall3Test:testAggregate3Value() (gas: 34736)
+Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62608)
 Multicall3Test:testAggregation() (gas: 5883)
```

Comment (abigger87) [reply src/Multicall3.sol:129]: Great catch! Added in #6 ! If you have any other suggestions feel free to fork and pr into this branch!

File: src/Multicall3.sol (hunk 1 above)
Comment (mds1) [reply src/Multicall3.sol]: I'd be ok with removing them also, and IIRC from Twitter I think @WyseNynja also supports removing them

Curious to hear arguments in favor of keeping these parameters as default return values. The main thing I can think of is that you probably always want at least one of block number, timestamp, or hash, and it's cheaper to include them by default compared to the cost of fetching them inside the `for` loop. But that cost difference is small compared to Geth's [default](https://github.com/ethereum/go-ethereum/blob/3038e480f5297b0e80196c156d9f9b45fd86a0bf/eth/ethconfig/config.go#L92) `RpcGasCap` of 50M and Infura's [default](https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_call) of 300M

Comment (abigger87) [reply src/Multicall3.sol]: Let's remove then; I think it makes sense to use exclusion by default here.
We should only be adding return values if necessary.

File: src/Multicall3.sol (hunk 6 above)
Comment (mds1) [reply src/Multicall3.sol]: done in d2320fb

File: src/Multicall3.sol (hunk 2 above)
Comment (mds1) [reply src/Multicall3.sol]: done in d2320fb

File: src/Multicall3.sol (hunk 4 above)
Comment (mds1) [reply src/Multicall3.sol]: Updated in d2320fb

**Review by abigger87 — APPROVED** [review]
👍

**Comment by mds1** [comment]
> Will moving the "Result memory result" (and any other variable declarations? outside the for loop save more gas? Assuming the stack isn't too deep, I think it will

Sorry what do you mean exactly? `result` holds the return data for the `i`th element in the array so `result` has to be assigned within the loop, unless I'm misunderstanding?

**Comment by WyseNynja** [comment]
> > Will moving the "Result memory result" (and any other variable declarations? outside the for loop save more gas? Assuming the stack isn't too deep, I think it will
> 
> Sorry what do you mean exactly? `result` holds the return data for the `i`th element in the array so `result` has to be assigned within the loop, unless I'm misunderstanding?

Assign it in the loop. But define the variable outside the loop.

**Comment by abigger87** [comment]
> > > Will moving the "Result memory result" (and any other variable declarations? outside the for loop save more gas? Assuming the stack isn't too deep, I think it will
> > 
> > 
> > Sorry what do you mean exactly? `result` holds the return data for the `i`th element in the array so `result` has to be assigned within the loop, unless I'm misunderstanding?
> 
> Assign it in the loop. But define the variable outside the loop. You already do this a few other places. Just not everywhere

Counterintuitively, it increases gas costs since it zero-initializes the memory I'm assuming (?)

Screenshot from 2022-03-08 10-22-00
